### PR TITLE
Improve activity input deserialization errors

### PR DESCRIPTION
**Pull request by YunchuWang**
# Summary
## What changed?
- Improve activity-input deserialization errors for **.NET isolated only**.
- Cover both isolated activity invocation paths:
  - Function-style activities through `ActivityInputConverter`, which throws the actionable `InvalidOperationException` directly when `DataConverter.Deserialize` fails.
  - Class-based `TaskActivity` implementations through `DurableFunctionExecutor`, which wraps failures from its direct `DataConverter.Deserialize` call in the same actionable exception.
- Identify the activity name and target type, explain that activity inputs must be concrete JSON-serializable values, and preserve the original serializer exception as the inner failure.

## Why is this change needed?
- Raw serializer failures do not explain the Durable Functions data boundary, which led users to try passing service instances or DI containers as activity inputs.
- The .NET in-process model is intentionally out of scope because it is approaching end of support.

## Compatibility impact
- For both .NET isolated function-style and class-based activities, input deserialization failures now use `System.InvalidOperationException` as the top-level `TaskFailureDetails.ErrorType`.
- The original serializer exception type and message are preserved in the structured `InnerFailure`. Consumers that need the serializer type should inspect `InnerFailure` instead of relying only on the top-level `ErrorType`.
- This behavior is based on the pinned Worker Core 2.51.0 pipeline: returned `ConversionResult.Failed` values are aggregated into a new `FunctionInputConverterException`, while an exception thrown directly by the converter propagates with its inner-exception chain intact.

## Issues / work items
- Related to #619
- Related to #3034

---

# Project checklist
- [x] Documentation changes are not required
  - [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
- [ ] Release notes are not required for the next release
  - [x] Otherwise: Notes added to `release_notes.md`
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] No extra work is required to be leveraged by OutOfProc SDKs
  - [ ] Otherwise: Work tracked here: #issue_or_pr_in_each_sdk
- [x] No change to the version of the `WebJobs.Extensions.DurableTask` package
  - [ ] Otherwise: Major/minor updates are reflected in `/src/Worker.Extensions.DurableTask/AssemblyInfo.cs`
- [x] No EventIds were added to `EventSource` logs
  - [ ] Otherwise: Ensure EventIds are within the supported range in the existing Windows infrastructure (validate via deployed telemetry). If needed, extend the range via a PR such as https://msazure.visualstudio.com/One/_git/AAPT-Antares-Websites/pullrequest/7463263?_a=files
- [ ] This change should be added to the `v2.x` branch
  - [x] Otherwise: This change applies only to `Microsoft.Azure.Functions.Worker.Extensions.DurableTask` (.NET isolated) and will be retained on the current development line
- [x] Breaking change? Yes (diagnostic compatibility)
  - [x] If yes:
    - Impact: `TaskFailureDetails.ErrorType` now reports `System.InvalidOperationException` for isolated activity input deserialization failures.
    - Migration guidance: Inspect `InnerFailure` for the original serializer exception type instead of relying only on the top-level `ErrorType`.
---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [x] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot CLI
- AI-assisted areas/files: implementation, unit tests, release notes, and PR description
- What you changed after AI output: Reviewed the final diff and focused test results.

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed
  - `dotnet test test\Worker.Extensions.DurableTask.Tests\Worker.Extensions.DurableTask.Tests.csproj --framework net8.0 --no-restore --filter "FullyQualifiedName~ActivityInputConverterTests|FullyQualifiedName~DurableFunctionExecutorActivityTests"`
  - `dotnet build src\Worker.Extensions.DurableTask\Worker.Extensions.DurableTask.csproj --no-restore`

## Manual validation (only if runtime/behavior changed)
- N/A; focused unit tests exercise both isolated deserialization paths and assert the actionable diagnostics, outer `TaskFailureDetails.ErrorType`, structured `InnerFailure`, and original exception instance.

---

# Notes for reviewers
- Each catch block surrounds only its corresponding `DataConverter.Deserialize` call, so the diagnostic wrapping is limited to deserialization failures.
- The function-style test verifies that `ActivityInputConverter.ConvertAsync` throws the actionable exception and that final failure serialization preserves the expected outer and inner error types.
- The class-based test exercises `DurableFunctionExecutor.ExecuteAsync` and validates the final serialized `TaskFailureDetails`, including the outer and inner error types.

### CODE REVIEW

File: src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs [hunk 1]
```diff
@@ -56,7 +56,19 @@ private async Task RunDirectActivityAsync(FunctionContext context, BindingMetada
...
 
-        object? input = this.Converter.Deserialize(data, activity.InputType);
+        object? input;
+        try
+        {
+            input = this.Converter.Deserialize(data, activity.InputType);
+        }
+        catch (Exception exception)
+        {
+            throw ActivityInputConverter.CreateDeserializationException(
```
Comment (YunchuWang) [inline src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs:66]: how do you know for sure this is deserialization failure?

File: src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs [hunk 2]
```diff
@@ -56,7 +56,19 @@ private async Task RunDirectActivityAsync(FunctionContext context, BindingMetada
...
+                context.FunctionDefinition.Name,
+                activity.InputType,
+                exception);
+        }
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs:70]: Addressed by the class-based executor test. It exercises DurableFunctionExecutor.ExecuteAsync and verifies the actionable message plus the final TaskFailureDetails outer and inner error types.

File: release_notes.md [hunk 3]
```diff
@@ -10,6 +10,7 @@
 ### Bug Fixes
 
 - Check if function invocation already has an executor before registering durable executor. (#3265)
+- Improved .NET isolated activity input deserialization diagnostics for both function-style and class-based activities. Failures now identify the activity and target type and preserve the original serializer error as the inner failure. `TaskFailureDetails.ErrorType` may now report `System.InvalidOperationException`; inspect `InnerFailure` for the original serializer exception type. (#3531)
 
 ### Breaking Changes
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply release_notes.md:16]: Fixed in 9668de49. The diagnostic improvement remains under Bug Fixes, and the top-level ErrorType change plus InnerFailure migration guidance now has a dedicated Breaking Changes entry.

File: test/Worker.Extensions.DurableTask.Tests/DurableFunctionExecutorActivityTests.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    private static IServiceProvider CreateServicesWithCachedInput(string input)
+    {
+        Assembly workerAssembly = typeof(FunctionContext).Assembly;
+        Type cacheInterface = workerAssembly.GetType("Microsoft.Azure.Functions.Worker.IBindingCache`1", true)!
+            .MakeGenericType(typeof(ConversionResult));
+        Type cacheImplementation = workerAssembly.GetType("Microsoft.Azure.Functions.Worker.DefaultBindingCache`1", true)!
+            .MakeGenericType(typeof(ConversionResult));
+        object cache = Activator.CreateInstance(cacheImplementation)!;
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply test/Worker.Extensions.DurableTask.Tests/DurableFunctionExecutorActivityTests.cs:90]: Fixed in 9668de49. Added a comment explaining that BindInputAsync requires Worker Core internal binding-cache state and that pre-population keeps this test focused on the executor deserialization path.

File: src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs (hunk 1 above)
Comment (YunchuWang) [reply src/Worker.Extensions.DurableTask/Execution/DurableFunctionExecutor.Activity.cs:66]: The catch block surrounds only this.Converter.Deserialize(data, activity.InputType). Activity lookup and input binding happen before the try, and activity execution happens after it, so only an exception thrown by the deserializer is wrapped with this diagnostic.

File: src/Worker.Extensions.DurableTask/ActivityInputConverter.cs [hunk 5]
```diff
@@ -47,7 +47,30 @@ public ValueTask<ConversionResult> ConvertAsync(ConverterContext context)
...
+        try
+        {
+            object? value = this.options.DataConverter.Deserialize(source, context.TargetType);
+            return new(ConversionResult.Success(value));
+        }
+        catch (Exception exception)
+        {
+            string activityName = context.FunctionContext.FunctionDefinition.Name;
+            return new(ConversionResult.Failed(
+                CreateDeserializationException(activityName, context.TargetType, exception)));
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply src/Worker.Extensions.DurableTask/ActivityInputConverter.cs]: Fixed in 9668de49. ActivityInputConverter now throws the actionable InvalidOperationException directly. Worker Core 2.51.0 does not catch converter-thrown exceptions in DefaultFunctionInputBindingFeature, so the chain reaches DurableFunctionExecutor intact. The updated test verifies the thrown exception, original exception identity, and serialized TaskFailureDetails with InvalidOperationException at the top and the serializer exception in InnerFailure.

**Comment by YunchuWang** [comment]
## Real Azure runtime validation

Confirmed against the real Azure Functions service using a disposable **Linux .NET 8 isolated Flex Consumption** app in `westus2` at PR head `9668de49d7f75ba0d0ee05740d7770cd2c241a77`.

The deployed health fingerprint reported:

```text
Durable extension: 1.19.0+9668de49d7f75ba0d0ee05740d7770cd2c241a77
Worker serializer: Azure.Core.Serialization.JsonObjectSerializer
Durable converter: Microsoft.Azure.Functions.Worker.Extensions.DurableTask.ObjectConverterShim
```

Both activities received `{"Count":"not-an-integer"}` for a DTO whose `Count` property is `int`. The `Failed` terminal status is expected: this test deliberately triggers input deserialization failure.

### Function-style `[ActivityTrigger]`

```text
Instance: 97c1c20418fc469b8be4624d419f3af7
RuntimeStatus: Failed
ErrorType: System.InvalidOperationException
ErrorMessage: Failed to deserialize input for activity function 'Pr3531FunctionStyleMismatchActivity' into type 'Microsoft.Azure.Durable.Tests.E2E.Pr3531MismatchDto'. Activity inputs must be JSON-serializable values. Pass concrete data transfer objects instead of interfaces or dependency-injected services, and ensure the input matches the target type. The original error was: The JSON value could not be converted to System.Int32. Path: $.Count | LineNumber: 0 | BytePositionInLine: 25.
InnerFailure.ErrorType: System.Text.Json.JsonException
InnerFailure.ErrorMessage: The JSON value could not be converted to System.Int32. Path: $.Count | LineNumber: 0 | BytePositionInLine: 25.
```

### Class-based `TaskActivity<TInput,TOutput>`

```text
Instance: f40c1ebe6f044e5faed9557d1676da5b
RuntimeStatus: Failed
ErrorType: System.InvalidOperationException
ErrorMessage: Failed to deserialize input for activity function 'Pr3531ClassBasedMismatchActivity' into type 'Microsoft.Azure.Durable.Tests.E2E.Pr3531MismatchDto'. Activity inputs must be JSON-serializable values. Pass concrete data transfer objects instead of interfaces or dependency-injected services, and ensure the input matches the target type. The original error was: The JSON value could not be converted to System.Int32. Path: $.Count | LineNumber: 0 | BytePositionInLine: 25.
InnerFailure.ErrorType: System.Text.Json.JsonException
InnerFailure.ErrorMessage: The JSON value could not be converted to System.Int32. Path: $.Count | LineNumber: 0 | BytePositionInLine: 25.
```

Both paths preserved the structured serializer inner failure and failed before the activity body executed.

@bachuv, could you please review this PR?

**Review by bachuv — APPROVED** [review]
